cli/flox-test-utils/src/lib.rs

```diff
@@ -0,0 +1,988 @@
+//   is the cause of some weird test failures (at cleanup time) in `bats`.
+// - By using `tempfile` we ensure that any temporary files we create are cleaned up
+//   when the test completes.
+// - `bats` creates a file for each test by concatenating the setup/body/teardown
+//   scripts, so you get a new tempfile for every test in the suite, every time you
+//   run it. Since this is a compiled artifact, you get one artifact for the entire
+//   suite.
+
+// Performance:
+// - Spawning a shell takes about 60ms
+// - `flox init` takes about 30ms
+// - Activating an empty environment takes 200-300ms
+// - By the time you've created dirs, done `flox init`, and activated, you're
+//   sitting pretty consistently around 350ms
+
+/// A collection of temporary directories to be used as an isolated home directory
+#[derive(Debug)]
+pub struct IsolatedHome {
+    _home_temp_dir: TempDir,
+    home_dir: PathBuf,
+    state_dir: PathBuf,
+    cache_dir: PathBuf,
+    data_dir: PathBuf,
+    config_dir: PathBuf,
+    envs: HashMap<String, String>,
+}
+
+impl IsolatedHome {
+    /// Returns true if the home dir and all of the XDG dirs that were originally created
+    /// still exist.
+    pub fn all_dirs_exist(&self) -> bool {
+        self.home_dir.exists()
+            && self.state_dir.exists()
+            && self.cache_dir.exists()
+            && self.data_dir.exists()
+            && self.config_dir.exists()
+    }
+}
+
+impl std::fmt::Display for IsolatedHome {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "IsolatedHome {{ ")?;
+        write!(
+            f,
+            "home_dir: {} (exists: {}), ",
+            self.home_dir.display(),
+            self.home_dir.exists()
+        )?;
+        write!(
+            f,
+            "state_dir: {} (exists: {}), ",
+            self.state_dir.display(),
+            self.state_dir.exists()
+        )?;
+        write!(
+            f,
+            "data_dir: {} (exists: {}), ",
+            self.data_dir.display(),
+            self.data_dir.exists()
+        )?;
+        write!(
+            f,
+            "config_dir: {} (exists: {}), ",
+            self.config_dir.display(),
+            self.config_dir.exists()
+        )?;
+        write!(
+            f,
+            "cache_dir: {} (exists: {}), ",
+            self.cache_dir.display(),
+            self.cache_dir.exists()
+        )?;
+        write!(f, " }}")?;
+
+        Ok(())
+    }
+}
+
+impl IsolatedHome {
+    pub fn new() -> Result<Self, Error> {
+        let home_tmp =
+            tempfile::TempDir::new().context("failed to create temporary home directory")?;
+        let home_dir = home_tmp.path().to_path_buf();
+        let data_dir = home_tmp.path().join(".local/share");
+        let state_dir = home_tmp.path().join(".local/state");
+        let config_dir = home_tmp.path().join(".config");
+        let cache_dir = home_tmp.path().join(".cache");
```

Comment:
nonblocking: how long is the path created by TempDir? See https://github.com/flox/flox/blob/b6c1dd7e0328327405d3b0555a49268fcc8c1d1f/cli/tests/setup_suite.bash#L377-L378

Reply:
I don't have an exact answer, but the path created here will be shorter than the one encountered in `bats` because there's at least one layer less of temporary directories (`bats` creates a tempdir for each run of the suite and puts stuff under it).